### PR TITLE
[enhance](Azure) Check delete operation's response on Azure Blob Storage

### PR DESCRIPTION
**Pull request by ByteYue**
We should check if the delete operation is successful or not.

### CODE REVIEW

File: be/src/io/fs/azure_obj_storage_client.cpp [hunk 1]
```diff
@@ -75,6 +78,40 @@ ObjectStorageResponse do_azure_client_call(Func f, const ObjectStoragePathOption
     return {};
 }
 
+struct AzureBatchDeleter {
```
Comment (gavinchou) [inline be/src/io/fs/azure_obj_storage_client.cpp:81]: type alias the name space for readability, such as using asb = Azure::Storage::Blobs

File: be/src/io/fs/azure_obj_storage_client.cpp [hunk 2]
```diff
@@ -75,6 +78,40 @@ ObjectStorageResponse do_azure_client_call(Func f, const ObjectStoragePathOption
...
+public:
+    AzureBatchDeleter(Azure::Storage::Blobs::BlobContainerClient* client,
```
Comment (gavinchou) [inline be/src/io/fs/azure_obj_storage_client.cpp:83]: add sufficient comments for all functions and params

File: be/src/io/fs/azure_obj_storage_client.cpp [hunk 3]
```diff
@@ -75,6 +78,40 @@ ObjectStorageResponse do_azure_client_call(Func f, const ObjectStoragePathOption
     return {};
 }
 
+struct AzureBatchDeleter {
+public:
```
Comment (gavinchou) [inline be/src/io/fs/azure_obj_storage_client.cpp:82]: redundant public?

File: be/src/io/fs/azure_obj_storage_client.cpp [hunk 4]
```diff
@@ -203,49 +240,59 @@ ObjectStorageResponse AzureObjStorageClient::delete_objects(const ObjectStorageP
...
-        auto batch = _client->CreateBatch();
+        auto deleter = AzureBatchDeleter(_client.get(), opts);
         auto chunkEnd = begin;
         std::advance(chunkEnd, std::min(BlobBatchMaxOperations,
                                         static_cast<size_t>(std::distance(begin, end))));
-        for (auto it = begin; it != chunkEnd; ++it) {
-            batch.DeleteBlob(*it);
-        }
+
+        std::ranges::for_each(std::ranges::subrange(begin, chunkEnd),
```
Comment (gavinchou) [inline be/src/io/fs/azure_obj_storage_client.cpp:248]: `chunkEnd` style is not correct, pls check other occurrences

and, capture only what is used

File: be/src/io/fs/azure_obj_storage_client.cpp [hunk 5]
```diff
@@ -75,6 +78,40 @@ ObjectStorageResponse do_azure_client_call(Func f, const ObjectStoragePathOption
...
+struct AzureBatchDeleter {
+public:
+    AzureBatchDeleter(Azure::Storage::Blobs::BlobContainerClient* client,
+                      const ObjectStoragePathOptions& opts)
+            : _client(client), _batch(client->CreateBatch()), _opts(opts) {}
+    void delete_blob(const std::string& blob_name) {
+        deferred_resps.emplace_back(_batch.DeleteBlob(blob_name));
+    }
+    ObjectStorageResponse execute() {
+        auto resp = do_azure_client_call([&]() { _client->SubmitBatch(_batch); }, _opts);
```
Comment (gavinchou) [inline be/src/io/fs/azure_obj_storage_client.cpp:90]: Is there any limit for this _batch, say is it ok that  10000000 blobs to delete?

**Review by gavinchou — CHANGES_REQUESTED** [review]

File: be/src/io/fs/azure_obj_storage_client.cpp [hunk 6]
```diff
@@ -75,6 +78,40 @@ ObjectStorageResponse do_azure_client_call(Func f, const ObjectStoragePathOption
...
+        if (std::ranges::any_of(
+                    deferred_resps,
+                    [](const Azure::Storage::DeferredResponse<
+                            Azure::Storage::Blobs::Models::DeleteBlobResult>& response) {
+                        return !response.GetResponse().Value.Deleted;
+                    })) {
+            return {.status = convert_to_obj_response(Status::InternalError<false>(
+                            "Failed to batch delete azure blob, path msg {}",
+                            wrap_object_storage_path_msg(_opts)))};
+        }
```
Comment (gavinchou) [inline be/src/io/fs/azure_obj_storage_client.cpp:103]:
```suggestion
        for (auto& i : deferred_resps) {
            if (i.GetResponse().Value.Deleted) continue;
            auto status = convert_to_obj_response(Status::InternalError<false>(
                                    "Failed to batch delete azure blob, path msg {}",
                                    wrap_object_storage_path_msg(_opts)))};
            return {.status = std::move(status)};
        }
```

**Review by gavinchou — CHANGES_REQUESTED** [review]

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

File: be/src/io/fs/azure_obj_storage_client.cpp [hunk 7]
```diff
@@ -78,6 +83,46 @@ ObjectStorageResponse do_azure_client_call(Func f, const ObjectStoragePathOption
...
+        };
+        for (auto&& defer_response : deferred_resps) {
+            auto response =
+                    do_azure_client_call([&]() { get_defer_response(defer_response); }, _opts);
+            if (response.status.code != ErrorCode::OK) {
+                return response;
+            }
+        }
+
+        return {};
```
Comment (gavinchou) [inline be/src/io/fs/azure_obj_storage_client.cpp:116]: explicitly set {statuscode = ok}

File: be/src/io/fs/azure_obj_storage_client.cpp [hunk 8]
```diff
@@ -210,49 +255,58 @@ ObjectStorageResponse AzureObjStorageClient::delete_objects(const ObjectStorageP
...
+    auto delete_func = [&](const std::vector<Models::BlobItem>& blobs) -> ObjectStorageResponse {
+        auto deleter = AzureBatchDeleter(_client.get(), opts);
+        auto batch = _client->CreateBatch();
+        for (auto&& blob_item : blobs) {
+            deleter.delete_blob(blob_item.Name);
+        }
+        if (auto response = deleter.execute(); response.status.code != ErrorCode::OK) {
+            return response;
+        }
+        return {};
```
Comment (gavinchou) [inline be/src/io/fs/azure_obj_storage_client.cpp:298]: ditto

File: cloud/src/recycler/azure_obj_client.cpp [hunk 9]
```diff
@@ -49,10 +52,57 @@ ObjectStorageResponse do_azure_client_call(Func f, const ObjectStoragePathOption
...
+
+        for (auto&& defer_response : deferred_resps) {
+            auto response =
+                    do_azure_client_call([&]() { get_defer_response(defer_response); }, _opts);
+            if (response.ret != 0) {
+                return response;
+            }
+        }
+
+        return {};
```
Comment (gavinchou) [inline cloud/src/recycler/azure_obj_client.cpp:96]: ditto

**Review by gavinchou — APPROVED** [review]

File: cloud/src/recycler/azure_obj_client.cpp (hunk 9 above)
Comment (ByteYue) [reply cloud/src/recycler/azure_obj_client.cpp:96]: I'll improve it in #36829

**Review by Yukang-Lian — APPROVED** [review]

**Pull request merged**